File: crates/ouros-js/session.ts
// Ergonomic TypeScript wrapper for the native SessionManager.
//
// Provides a `SessionManager` class that creates `Session` objects, each
// representing a named interpreter session with methods for execution,
// variable management, history/rewind, and heap inspection.

import type { NapiExecuteResult, NapiVariableValue, NapiEvalResult, NapiHeapDiffResult } from './index.js'

import { SessionManager as NativeSessionManager } from './index.js'

// =============================================================================
// Result types - ergonomic wrappers around napi output types
// =============================================================================

/**
 * Result of executing code in a session.
 *
 * Contains execution output (stdout, result value, completion status)
 * and progress state when paused at an external function call.
 */
export interface ExecuteResult {
  /** Standard output captured during execution. */
  stdout: string
  /** Whether execution completed (true) or paused at an external call (false). */
  isComplete: boolean
  /** The result value, parsed from JSON. `undefined` for None or non-complete progress. */
  result: unknown
  /** For FunctionCall progress: the external function name. */
  functionName?: string
  /** For FunctionCall progress: the call ID for resume. */
  callId?: number
  /** For FunctionCall progress: positional args parsed from JSON. */
  args?: unknown[]
  /** For ResolveFutures progress: pending call IDs. */
  pendingCallIds?: number[]
}

/**
 * Info about a single variable in a session.
 */
export interface VariableInfo {
  /** Variable name. */
  name: string
  /** Python type name (e.g. "int", "str", "list"). */
  typeName: string
}

/**
 * A variable's value with JSON and repr representations.
 */
export interface VariableValue {
  /** The value, parsed from JSON. */
  jsonValue: unknown
  /** Python `repr()` string, if available. */
  repr: string | null
}

/**
 * Result of evaluating an expression without side effects.
 */
export interface EvalResult {
  /** The computed value. */
  value: VariableValue
  /** Standard output captured during evaluation. */
  stdout: string
}

/**
 * Summary info for one active session.
 */
export interface SessionInfo {
  /** The session ID. */
  id: string
  /** Number of defined global variables. */
  variableCount: number
}

/**
 * Result of a rewind operation.
 */
export interface RewindResult {
  /** Number of steps actually rewound. */
  stepsRewound: number
  /** Number of history entries remaining after rewind. */
  historyRemaining: number
}

/**
 * History depth info for a session.
 */
export interface HistoryInfo {
  /** Current undo history depth. */
  current: number
  /** Maximum configured history depth. */
  max: number
}

/**
 * Heap statistics for a session.
 */
export interface HeapStats {
  /** Total number of live objects on the heap. */
  liveObjects: number
  /** Number of free (recycled) slots. */
  freeSlots: number
  /** Total heap capacity (live + free). */
  totalSlots: number
  /** Number of interned strings. */
  internedStrings: number
}

/**
 * Aggregate heap counter deltas between two snapshots.
 */
export interface HeapDiff {
  /** Change in live object count. */
  liveObjectsDelta: number
  /** Change in free slot count. */
  freeSlotsDelta: number
  /** Change in total slot count. */
  totalSlotsDelta: number
  /** Change in interned string count. */
  internedStringsDelta: number
}

/**
 * One variable that changed between two heap snapshots.
 */
export interface ChangedVariable {
  /** Variable name. */
  name: string
  /** Repr string in the "before" state. */
  before: string
  /** Repr string in the "after" state. */
  after: string
}

/**
 * Variable-level diff summary between two heap states.
 */
export interface VariableDiff {
  /** Variables present only in the "after" state. */
  added: string[]
  /** Variables present only in the "before" state. */
  removed: string[]
  /** Variables whose repr changed between states. */
  changed: ChangedVariable[]
  /** Variables with identical repr in both states. */
  unchanged: string[]
}

/**
 * Result of a heap diff operation.
 */
export interface HeapDiffResult {
  /** Aggregate heap counter deltas. */
  heapDiff: HeapDiff
  /** Variable-level changes between two states. */
  variableDiff: VariableDiff
}

/**
 * Result of a save operation.
 */
export interface SaveResult {
  /** The snapshot name used for the save. */
  name: string
  /** Size of the saved snapshot in bytes. */
  sizeBytes: number
}

/**
 * Info about a saved session snapshot on disk.
 */
export interface SavedSessionInfo {
  /** Snapshot name (filename without extension). */
  name: string
  /** Size of the snapshot file in bytes. */
  sizeBytes: number
}

/**
 * Options for creating a new SessionManager.
 */
export interface SessionManagerOptions {
  /** Script name for tracebacks and error messages. Defaults to "session.py". */
  scriptName?: string
  /** Directory for session persistence (save/load). If not set, persistence is disabled. */
  storageDir?: string
}

// =============================================================================
// Conversion helpers
// =============================================================================

/** Parses the napi execute result into an ergonomic `ExecuteResult`. */
function wrapExecuteResult(raw: NapiExecuteResult): ExecuteResult {
  let result: unknown = undefined
  if (raw.resultJson != null) {
    try {
      result = JSON.parse(raw.resultJson)
    } catch {
      result = raw.resultJson
    }
  }

  let args: unknown[] | undefined = undefined
  if (raw.argsJson != null) {
    try {
      args = JSON.parse(raw.argsJson) as unknown[]
    } catch {
      args = undefined
    }
  }

  return {
    stdout: raw.stdout,
    isComplete: raw.isComplete,
    result,
    functionName: raw.functionName ?? undefined,
    callId: raw.callId ?? undefined,
    args,
    pendingCallIds: raw.pendingCallIds ?? undefined,
  }
}

/** Converts a napi variable value (with JSON string) to the ergonomic form. */
function wrapVariableValue(raw: NapiVariableValue): VariableValue {
  let jsonValue: unknown
  try {
    jsonValue = JSON.parse(raw.jsonValue)
  } catch {
    jsonValue = raw.jsonValue
  }
  return {
    jsonValue,
    repr: raw.repr ?? null,
  }
}

/** Converts a napi eval result to the ergonomic form. */
function wrapEvalResult(raw: NapiEvalResult): EvalResult {
  return {
    value: wrapVariableValue(raw.value),
    stdout: raw.stdout,
  }
}

/** Converts a napi heap diff result to the ergonomic form. */
function wrapHeapDiffResult(raw: NapiHeapDiffResult): HeapDiffResult {
  return {
    heapDiff: {
      liveObjectsDelta: raw.heapDiff.liveObjectsDelta,
      freeSlotsDelta: raw.heapDiff.freeSlotsDelta,
      totalSlotsDelta: raw.heapDiff.totalSlotsDelta,
      internedStringsDelta: raw.heapDiff.internedStringsDelta,
    },
    variableDiff: {
      added: raw.variableDiff.added,
      removed: raw.variableDiff.removed,
      changed: raw.variableDiff.changed.map((c) => ({
        name: c.name,
        before: c.before,
        after: c.after,
      })),
      unchanged: raw.variableDiff.unchanged,
    },
  }
}

// =============================================================================
// Session - A handle to one named session within a SessionManager
// =============================================================================

/**
 * Represents a single named interpreter session within a `SessionManager`.
 *
 * This is a lightweight handle that delegates all operations to the underlying
 * native `SessionManager`, tagging each call with this session's ID.
 *
 * Do not construct directly -- use `SessionManager.createSession()` or
 * `SessionManager.getSession()`.
 */
export class Session {
  private _manager: SessionManager
  private _id: string

  /** @internal */
  constructor(manager: SessionManager, sessionId: string) {
    this._manager = manager
    this._id = sessionId
  }

  /** The session ID. */
  get id(): string {
    return this._id
  }

  // -------------------------------------------------------------------------
  // Execution
  // -------------------------------------------------------------------------

  /**
   * Executes Python code in this session.
   *
   * @param code - Python code to execute
   * @returns Execution result with stdout, completion status, and value
   * @throws When the code has syntax errors or raises a runtime exception
   */
  execute(code: string): ExecuteResult {
    return wrapExecuteResult(this._manager._native.execute(code, this._id))
  }

  // -------------------------------------------------------------------------
  // Variables
  // -------------------------------------------------------------------------

  /**
   * Lists defined global variables and their types.
   */
  listVariables(): VariableInfo[] {
    return this._manager._native.listVariables(this._id).map((v) => ({
      name: v.name,
      typeName: v.typeName,
    }))
  }

  /**
   * Gets one variable's value from the session namespace.
   *
   * @param name - Variable name
   * @returns The variable value with JSON and repr representations
   * @throws When the variable does not exist
   */
  getVariable(name: string): VariableValue {
    return wrapVariableValue(this._manager._native.getVariable(this._id, name))
  }

  /**
   * Sets or creates a global variable by evaluating a Python expression.
   *
   * @param name - Variable name to set
   * @param valueExpr - Python expression to evaluate (e.g. "[1, 2, 3]", "'hello'")
   */
  setVariable(name: string, valueExpr: string): void {
    this._manager._native.setVariable(this._id, name, valueExpr)
  }

  /**
   * Deletes a global variable from the session.
   *
   * @param name - Variable name to delete
   * @returns true if the variable existed and was removed, false otherwise
   */
  deleteVariable(name: string): boolean {
    return this._manager._native.deleteVariable(this._id, name)
  }

  /**
   * Evaluates a Python expression without modifying session state.
   *
   * The expression is executed in a forked copy of the session, so no
   * side effects are persisted.
   *
   * @param expression - Python expression to evaluate
   * @returns The computed value and any captured stdout
   */
  evalVariable(expression: string): EvalResult {
    return wrapEvalResult(this._manager._native.evalVariable(this._id, expression))
  }

  // -------------------------------------------------------------------------
  // Fork
  // -------------------------------------------------------------------------

  /**
   * Forks this session into a new independent copy.
   *
   * The forked session starts with the same state (variables, heap) but
   * subsequent modifications are independent.
   *
   * @param newId - ID for the forked session
   * @returns A new Session handle for the fork
   */
  fork(newId: string): Session {
    this._manager._native.forkSession(this._id, newId)
    return new Session(this._manager, newId)
  }

  // -------------------------------------------------------------------------
  // History / rewind
  // -------------------------------------------------------------------------

  /**
   * Rewinds this session by N steps, restoring a previous state.
   *
   * @param steps - Number of steps to rewind (default: 1)
   * @returns Information about the rewind (steps rewound, remaining history)
   */
  rewind(steps: number = 1): RewindResult {
    const raw = this._manager._native.rewind(this._id, steps)
    return { stepsRewound: raw.stepsRewound, historyRemaining: raw.historyRemaining }
  }

  /**
   * Returns the current undo history depth and configured maximum.
   */
  history(): HistoryInfo {
    const raw = this._manager._native.history(this._id)
    return { current: raw.current, max: raw.max }
  }

  /**
   * Configures the maximum undo history depth for this session.
   *
   * If the new maximum is less than the current depth, the oldest entries
   * are trimmed. Returns the number of entries that were trimmed.
   *
   * @param maxDepth - New maximum history depth
   * @returns Number of entries that were trimmed
   */
  setHistoryDepth(maxDepth: number): number {
    return this._manager._native.setHistoryDepth(this._id, maxDepth)
  }

  // -------------------------------------------------------------------------
  // Heap introspection
  // -------------------------------------------------------------------------

  /**
   * Returns heap statistics for this session.
   */
  heapStats(): HeapStats {
    const raw = this._manager._native.heapStats(this._id)
    return {
      liveObjects: raw.liveObjects,
      freeSlots: raw.freeSlots,
      totalSlots: raw.totalSlots,
      internedStrings: raw.internedStrings,
    }
  }

  // -------------------------------------------------------------------------
  // Reset
  // -------------------------------------------------------------------------

  /**
   * Resets this session, replacing it with a fresh interpreter instance.
   *
   * The session's history is cleared. Pass external function names if the
   * session needs them after reset.
   *
   * @param externalFunctions - External function names for the reset session
   */
  reset(externalFunctions?: string[]): void {
    this._manager._native.reset(this._id, externalFunctions)
  }
}

// =============================================================================
// SessionManager - Multi-session interpreter manager
// =============================================================================

/**
 * Multi-session Python interpreter manager.
 *
 * Manages a registry of named interpreter sessions, each with its own
 * variables, heap, and execution history. A "default" session is always
 * present and is used when no session ID is specified.
 *
 * @example
 * ```typescript
 * const mgr = new SessionManager()
 * const session = mgr.getSession()
 * session.execute('x = 42')
 * console.log(session.getVariable('x').jsonValue) // 42
 *
 * const other = mgr.createSession('analysis')
 * other.execute('data = [1, 2, 3]')
 * mgr.transferVariable('analysis', 'default', 'data')
 * ```
 */
export class SessionManager {
  /** @internal -- the native napi session manager instance. */
  _native: NativeSessionManager

  /**
   * Creates a new session manager.
   *
   * @param options - Configuration options (script name, storage directory)
   */
  constructor(options?: SessionManagerOptions) {
    this._native = new NativeSessionManager(options?.scriptName, options?.storageDir)
  }

  // -------------------------------------------------------------------------
  // Session lifecycle
  // -------------------------------------------------------------------------

  /**
   * Creates a new named session.
   *
   * @param sessionId - Unique ID for the new session
   * @param options - Optional configuration (external function names)
   * @returns A Session handle for the new session
   * @throws When a session with the given ID already exists
   */
  createSession(sessionId: string, options?: { externalFunctions?: string[] }): Session {
    this._native.createSession(sessionId, options?.externalFunctions)
    return new Session(this, sessionId)
  }

  /**
   * Gets a Session handle for an existing session.
   *
   * @param sessionId - Session ID (default: the "default" session)
   * @returns A Session handle
   */
  getSession(sessionId?: string): Session {
    return new Session(this, sessionId ?? 'default')
  }

  /**
   * Destroys a named session. The default session cannot be destroyed.
   *
   * @param sessionId - ID of the session to destroy
   * @throws When the session is the default or does not exist
   */
  destroySession(sessionId: string): void {
    this._native.destroySession(sessionId)
  }

  /**
   * Lists all active sessions with their variable counts.
   */
  listSessions(): SessionInfo[] {
    return this._native.listSessions().map((s) => ({
      id: s.id,
      variableCount: s.variableCount,
    }))
  }

  // -------------------------------------------------------------------------
  // Cross-session operations
  // -------------------------------------------------------------------------

  /**
   * Transfers a variable from one session to another.
   *
   * Reads the variable from the source session and writes it into the
   * target session. The transfer is heap-independent (no raw references leak).
   *
   * @param source - Source session ID
   * @param target - Target session ID
   * @param name - Variable name in the source session
   * @param targetName - Optional different name in the target session
   */
  transferVariable(source: string, target: string, name: string, targetName?: string): void {
    this._native.transferVariable(source, target, name, targetName)
  }

  /**
   * Executes code in a source session and stores the result in a target session.
   *
   * @param source - Source session ID (null for default)
   * @param target - Target session ID
   * @param code - Python code to execute in the source
   * @param targetVariable - Variable name in the target to store the result
   * @returns Execution result
   */
  callSession(source: string | null, target: string, code: string, targetVariable: string): ExecuteResult {
    return wrapExecuteResult(this._native.callSession(source, target, code, targetVariable))
  }

  // -------------------------------------------------------------------------
  // Heap introspection
  // -------------------------------------------------------------------------

  /**
   * Saves the current heap stats as a named snapshot for later diff.
   *
   * @param name - Snapshot name
   * @param sessionId - Session ID (default: the "default" session)
   */
  snapshotHeap(name: string, sessionId?: string): void {
    this._native.snapshotHeap(sessionId ?? null, name)
  }

  /**
   * Compares two named heap snapshots and returns the diff.
   *
   * @param before - Name of the "before" snapshot
   * @param after - Name of the "after" snapshot
   * @returns Heap diff with aggregate deltas and variable-level changes
   */
  diffHeap(before: string, after: string): HeapDiffResult {
    return wrapHeapDiffResult(this._native.diffHeap(before, after))
  }

  // -------------------------------------------------------------------------
  // Persistence
  // -------------------------------------------------------------------------

  /**
   * Saves a session to disk as a named snapshot.
   *
   * Requires `storageDir` to be set in the constructor options.
   *
   * @param sessionId - Session ID (default: the "default" session)
   * @param name - Snapshot name (default: the session ID)
   * @returns Save result with name and size
   */
  saveSession(sessionId?: string, name?: string): SaveResult {
    const raw = this._native.saveSession(sessionId, name)
    return { name: raw.name, sizeBytes: raw.sizeBytes }
  }

  /**
   * Loads a previously saved session from disk.
   *
   * @param name - Snapshot name to load
   * @param sessionId - ID for the new session (default: the snapshot name)
   * @returns The session ID that was created
   */
  loadSession(name: string, sessionId?: string): string {
    return this._native.loadSession(name, sessionId)
  }

  /**
   * Lists all saved session snapshots on disk.
   */
  listSavedSessions(): SavedSessionInfo[] {
    return this._native.listSavedSessions().map((s) => ({
      name: s.name,
      sizeBytes: s.sizeBytes,
    }))
  }
}

File: crates/ouros-python/python/ouros/session.py
"""Multi-session interpreter management for Ouros.

Provides :class:`SessionManager` and :class:`Session` as ergonomic Python
wrappers around the native Rust ``_ouros.SessionManager`` class.

:class:`Session` is a thin convenience handle that binds a ``session_id`` to
a manager, letting callers write ``session.execute(code)`` instead of
``manager.execute(code, session_id='...')``.

Example::

    from ouros.session import SessionManager

    mgr = SessionManager()
    s = mgr.create_session('sandbox')
    s.execute('x = 42')
    print(s.get_variable('x'))  # {'json_value': 42, 'repr': '42'}
"""

from __future__ import annotations

from typing import Any

from ._ouros import SessionManager as _NativeSessionManager


class Session:
    """A single interpreter session bound to a :class:`SessionManager`.

    All operations delegate to the underlying native manager, passing
    ``session_id`` automatically. Do not construct directly -- use
    :meth:`SessionManager.create_session` instead.
    """

    __slots__ = ('_manager', '_id')

    def __init__(self, manager: SessionManager, session_id: str) -> None:
        self._manager = manager
        self._id = session_id

    @property
    def id(self) -> str:
        """The session ID string."""
        return self._id

    # -- Execution -------------------------------------------------------------

    def execute(self, code: str) -> dict[str, Any]:
        """Execute Python code in this session."""
        return self._manager._native.execute(code, session_id=self._id)

    def resume(self, call_id: int, value: Any) -> dict[str, Any]:
        """Resume execution after an external function call."""
        return self._manager._native.resume(call_id, value, session_id=self._id)

    def resume_as_pending(self, call_id: int) -> dict[str, Any]:
        """Resume an external call as a pending async future."""
        return self._manager._native.resume_as_pending(call_id, session_id=self._id)

    def resume_futures(self, results: dict[int, Any]) -> dict[str, Any]:
        """Resume execution with results for pending futures."""
        return self._manager._native.resume_futures(results, session_id=self._id)

    # -- Variables -------------------------------------------------------------

    def list_variables(self) -> list[dict[str, Any]]:
        """List defined global variables and their types."""
        return self._manager._native.list_variables(session_id=self._id)

    def get_variable(self, name: str) -> dict[str, Any]:
        """Get one variable's value."""
        return self._manager._native.get_variable(name, session_id=self._id)

    def set_variable(self, name: str, value_expr: str) -> None:
        """Set or create a global variable via a Python expression string."""
        self._manager._native.set_variable(name, value_expr, session_id=self._id)

    def delete_variable(self, name: str) -> bool:
        """Delete a global variable. Returns True if it existed."""
        return self._manager._native.delete_variable(name, session_id=self._id)

    def eval_variable(self, expression: str) -> dict[str, Any]:
        """Evaluate a Python expression without modifying session state."""
        return self._manager._native.eval_variable(expression, session_id=self._id)

    # -- Fork ------------------------------------------------------------------

    def fork(self, new_id: str) -> Session:
        """Fork this session into a new independent copy."""
        self._manager._native.fork_session(self._id, new_id)
        return Session(self._manager, new_id)

    # -- History / rewind ------------------------------------------------------

    def rewind(self, *, steps: int = 1) -> dict[str, Any]:
        """Rewind this session by N steps."""
        return self._manager._native.rewind(steps=steps, session_id=self._id)

    def history(self) -> tuple[int, int]:
        """Returns (current_depth, max_depth) for session history."""
        return self._manager._native.history(session_id=self._id)

    def set_history_depth(self, max_depth: int) -> int:
        """Configure the maximum undo history depth. Returns entries trimmed."""
        return self._manager._native.set_history_depth(max_depth, session_id=self._id)

    # -- Heap ------------------------------------------------------------------

    def heap_stats(self) -> dict[str, Any]:
        """Return heap statistics for this session."""
        return self._manager._native.heap_stats(session_id=self._id)

    def snapshot_heap(self, name: str) -> None:
        """Save the current heap stats as a named snapshot."""
        self._manager._native.snapshot_heap(name, session_id=self._id)

    # -- Persistence -----------------------------------------------------------

    def save(self, *, name: str | None = None) -> dict[str, Any]:
        """Save this session to disk."""
        return self._manager._native.save_session(session_id=self._id, name=name)

    # -- Reset -----------------------------------------------------------------

    def reset(self, *, external_functions: list[str] | None = None) -> None:
        """Reset this session to a fresh state."""
        self._manager._native.reset(session_id=self._id, external_functions=external_functions)

    def __repr__(self) -> str:
        return f'Session(id={self._id!r})'


class SessionManager:
    """Multi-session manager for the Ouros interpreter.

    Wraps the native Rust ``SessionManager`` and provides a Pythonic API.
    A "default" session is always present and is used when no ``session_id``
    is specified.
    """

    __slots__ = ('_native',)

    def __init__(self, *, script_name: str = 'session.py') -> None:
        self._native = _NativeSessionManager(script_name=script_name)

    # -- Session lifecycle -----------------------------------------------------

    def create_session(
        self,
        session_id: str,
        *,
        external_functions: list[str] | None = None,
    ) -> Session:
        """Create a new named session and return a Session handle."""
        self._native.create_session(session_id, external_functions=external_functions)
        return Session(self, session_id)

    def destroy_session(self, session_id: str) -> None:
        """Destroy a named session."""
        self._native.destroy_session(session_id)

    def list_sessions(self) -> list[dict[str, Any]]:
        """List all active sessions with their variable counts."""
        return self._native.list_sessions()

    def fork_session(self, source: str, new_id: str) -> Session:
        """Fork an existing session into a new independent copy."""
        self._native.fork_session(source, new_id)
        return Session(self, new_id)

    # -- Execution (delegate to default session) -------------------------------

    def execute(self, code: str, *, session_id: str | None = None) -> dict[str, Any]:
        """Execute Python code in the default or specified session."""
        return self._native.execute(code, session_id=session_id)

    # -- Variables (delegate to default session) -------------------------------

    def list_variables(self, *, session_id: str | None = None) -> list[dict[str, Any]]:
        """List defined global variables."""
        return self._native.list_variables(session_id=session_id)

    def get_variable(self, name: str, *, session_id: str | None = None) -> dict[str, Any]:
        """Get one variable's value."""
        return self._native.get_variable(name, session_id=session_id)

    def set_variable(self, name: str, value_expr: str, *, session_id: str | None = None) -> None:
        """Set or create a global variable via a Python expression string."""
        self._native.set_variable(name, value_expr, session_id=session_id)

    def delete_variable(self, name: str, *, session_id: str | None = None) -> bool:
        """Delete a global variable. Returns True if it existed."""
        return self._native.delete_variable(name, session_id=session_id)

    def eval_variable(self, expression: str, *, session_id: str | None = None) -> dict[str, Any]:
        """Evaluate a Python expression without modifying session state."""
        return self._native.eval_variable(expression, session_id=session_id)

    def transfer_variable(
        self,
        source: str,
        target: str,
        name: str,
        *,
        target_name: str | None = None,
    ) -> None:
        """Transfer a variable from one session to another."""
        self._native.transfer_variable(source, target, name, target_name=target_name)

    # -- History / rewind (delegate to default session) ------------------------

    def rewind(self, *, steps: int = 1, session_id: str | None = None) -> dict[str, Any]:
        """Rewind a session by N steps."""
        return self._native.rewind(steps=steps, session_id=session_id)

    def history(self, *, session_id: str | None = None) -> tuple[int, int]:
        """Returns (current_depth, max_depth) for session history."""
        return self._native.history(session_id=session_id)

    def set_history_depth(self, max_depth: int, *, session_id: str | None = None) -> int:
        """Configure the maximum undo history depth. Returns entries trimmed."""
        return self._native.set_history_depth(max_depth, session_id=session_id)

    # -- Heap introspection (delegate to default session) ----------------------

    def heap_stats(self, *, session_id: str | None = None) -> dict[str, Any]:
        """Return heap statistics for a session."""
        return self._native.heap_stats(session_id=session_id)

    def snapshot_heap(self, name: str, *, session_id: str | None = None) -> None:
        """Save the current heap stats as a named snapshot."""
        self._native.snapshot_heap(name, session_id=session_id)

    def diff_heap(self, before: str, after: str) -> dict[str, Any]:
        """Compare two named heap snapshots and return the diff."""
        return self._native.diff_heap(before, after)

    # -- Persistence -----------------------------------------------------------

    def set_storage_dir(self, directory: str) -> None:
        """Configure the directory for session persistence."""
        self._native.set_storage_dir(directory)

    def save_session(
        self,
        *,
        session_id: str | None = None,
        name: str | None = None,
    ) -> dict[str, Any]:
        """Save a session to disk as a named snapshot."""
        return self._native.save_session(session_id=session_id, name=name)

    def load_session(self, name: str, *, session_id: str | None = None) -> str:
        """Load a previously saved session from disk. Returns the session ID."""
        return self._native.load_session(name, session_id=session_id)

    def list_saved_sessions(self) -> list[dict[str, Any]]:
        """List all saved session snapshots on disk."""
        return self._native.list_saved_sessions()

    # -- Reset -----------------------------------------------------------------

    def reset(self, *, session_id: str | None = None, external_functions: list[str] | None = None) -> None:
        """Reset a session to a fresh state."""
        self._native.reset(session_id=session_id, external_functions=external_functions)

    # -- Cross-session pipeline ------------------------------------------------

    def call_session(
        self,
        *,
        target: str,
        code: str,
        target_variable: str,
        source: str | None = None,
    ) -> dict[str, Any]:
        """Execute code in source session and store result in target session."""
        return self._native.call_session(
            target=target,
            code=code,
            target_variable=target_variable,
            source=source,
        )

    def __repr__(self) -> str:
        return 'SessionManager()'
